<img src="https://d3c33hcgiwev3.cloudfront.net/imageAssetProxy.v1/xCm_NX3ARhGby5yNHxROZg_2a48fa4727ec4aaf9bd7e063838687f1_image.png?expiry=1738540800000&hmac=pVRbB3GMVL2gz_fozxTWFPYcuYRwl5e1XBccQJJeDww" alt="Description" style="width:70%; height:auto; float: left;">

<font color='#4393C3' size='4'><b>Project goal:</b></font> Waze leadership has asked your data team to develop a machine learning model to predict user churn. Churn quantifies the number of users who have uninstalled the Waze app or stopped using the app. This project focuses on monthly user churn. An accurate model will help prevent churn, improve user retention, and grow Waze’s business.
This project is part of a larger effort at Waze to increase growth. Typically, high retention rates indicate satisfied users who repeatedly use the Waze app over time. Developing a churn prediction model will help prevent churn, improve user retention, and grow Waze’s business. If Waze can identify a segment of users who are at high risk of churning, Waze can proactively engage these users with special offers to try and retain them. Otherwise, Waze may lose these users without knowing why. 
Your insights will help Waze leadership optimize the company’s retention strategy, enhance user experience, and make data-driven decisions about product development.

<font color='#4393C3' size='4'><b>Background:</b></font> Waze’s free navigation app makes it easier for drivers around the world to get to where they want to go. Waze’s community of map editors, beta testers, translators, partners, and users helps make each drive better and safer. 

<font color='#4393C3' size='4'><b>Scenario:</b></font> Your team is in the early stages of their user churn project. Your project proposal has been approved and your team has been given access to Waze’s user data. To get clear insights, the data must first be inspected, organized, and prepared for analysis. 

To get clear insights, the data must first be <ins>inspected, organized, and prepared for analysis</ins>. 

<font color='#4393C3' size='4'><b>Tasks:</b></font> 
- Import data and build a dataframe 
- Examine data 
- Identify outliers
- Gather descriptive statistics
- Create a data visualization
- Prepare executive summary 

<font color='#4393C3' size='5' style="text-decoration: underline; text-decoration-color: blue;"><b>Import the Packages and Load The Data</b></font>

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from scipy import stats
import statsmodels.api as sm

In [2]:
waze_data = pd.read_csv("/kaggle/input/waze-datasetgoogle-advanced-data-analytics/waze_dataset(Google Advanced Data Analytics).csv")

<font color='#4393C3' size='5' style="text-decoration: underline; text-decoration-color: blue;"><b>Dataset Summary</b></font>

In [3]:
waze_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14999 entries, 0 to 14998
Data columns (total 13 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   ID                       14999 non-null  int64  
 1   label                    14299 non-null  object 
 2   sessions                 14999 non-null  int64  
 3   drives                   14999 non-null  int64  
 4   total_sessions           14999 non-null  float64
 5   n_days_after_onboarding  14999 non-null  int64  
 6   total_navigations_fav1   14999 non-null  int64  
 7   total_navigations_fav2   14999 non-null  int64  
 8   driven_km_drives         14999 non-null  float64
 9   duration_minutes_drives  14999 non-null  float64
 10  activity_days            14999 non-null  int64  
 11  driving_days             14999 non-null  int64  
 12  device                   14999 non-null  object 
dtypes: float64(3), int64(8), object(2)
memory usage: 1.5+ MB


**Dataset Description**: This dataset contains information about user engagement and driving data. It has **14999** rows (users) and **13** columns.

**Feature Descriptions**: 
- **ID** (a sequential numbered index)
- **label** (binary  variable: “retained” vs “churned”)
- **sessions** (number of times a user opened the app during the month)
- **drives** (number of times a drive longer than 1 km was taken during the month)
- **total_sessions** (a total number of sessions registered for a user)
- **n_days_after_onboarding** (the number of days since a user signed up for the app)
- **total_navigations_fav1** (total navigations since onboarding to the user’s favorite place 1)
- **total_navigations_fav2** (total navigations since onboarding to the user’s favorite place 2)
- **driven_km_drives** (total kilometers driven during the month)
- **duration_minutes_drives** (total duration driven in minutes during the month)
- **activity_days** (number of days a user opens the app during the month)
- **driving_days** (number of times a drive longer than 1 km was taken during the month)
- **device** (type of device: iPhone or Android)

In [4]:
waze_data['label'].isnull().value_counts().reset_index()

,label,count
0,False,14299
1,True,700


The total number of entries is 14999. The column 'label' has only 14299 non-null values. Since the column with the missing data contains information about the status of a user (retained or churned) we cannot take the rows with NaN into consideration during analysis (we either should try to find the missing information or perform the analysis without rows with missing data).

In [5]:
waze_data.head(100)

,ID,label,sessions,drives,total_sessions,n_days_after_onboarding,total_navigations_fav1,total_navigations_fav2,driven_km_drives,duration_minutes_drives,activity_days,driving_days,device
0,0,retained,283,226,296.748273,2276,208,0,2628.845068,1985.775061,28,19,Android
1,1,retained,133,107,326.896596,1225,19,64,13715.920550,3160.472914,13,11,iPhone
2,2,retained,114,95,135.522926,2651,0,0,3059.148818,1610.735904,14,8,Android
3,3,retained,49,40,67.589221,15,322,7,913.591123,587.196542,7,3,iPhone
4,4,retained,84,68,168.247020,1562,166,5,3950.202008,1219.555924,27,18,Android
...,...,...,...,...,...,...,...,...,...,...,...,...,...
95,95,churned,116,95,192.638291,359,142,0,4032.868589,2695.612209,1,0,iPhone
96,96,churned,305,244,638.040152,400,171,164,1896.894034,960.975077,19,13,Android
97,97,retained,0,0,559.547112,2879,0,0,6668.844350,3511.370419,28,24,Android
98,98,NaN,78,64,583.492789,3414,0,0,1811.140893,642.189122,12,11,Android


In [6]:
waze_data.describe()

,ID,sessions,drives,total_sessions,n_days_after_onboarding,total_navigations_fav1,total_navigations_fav2,driven_km_drives,duration_minutes_drives,activity_days,driving_days
count,14999.000000,14999.000000,14999.000000,14999.000000,14999.000000,14999.000000,14999.000000,14999.000000,14999.000000,14999.000000,14999.000000
mean,7499.000000,80.633776,67.281152,189.964447,1749.837789,121.605974,29.672512,4039.340921,1860.976012,15.537102,12.179879
std,4329.982679,80.699065,65.913872,136.405128,1008.513876,148.121544,45.394651,2502.149334,1446.702288,9.004655,7.824036
min,0.000000,0.000000,0.000000,0.220211,4.000000,0.000000,0.000000,60.441250,18.282082,0.000000,0.000000
25%,3749.500000,23.000000,20.000000,90.661156,878.000000,9.000000,0.000000,2212.600607,835.996260,8.000000,5.000000
50%,7499.000000,56.000000,48.000000,159.568115,1741.000000,71.000000,9.000000,3493.858085,1478.249859,16.000000,12.000000
75%,11248.500000,112.000000,93.000000,254.192341,2623.500000,178.000000,43.000000,5289.861262,2464.362632,23.000000,19.000000
max,14998.000000,743.000000,596.000000,1216.154633,3500.000000,1236.000000,415.000000,21183.401890,15851.727160,31.000000,30.000000


<font color='#4393C3' size='5' style="text-decoration: underline; text-decoration-color: blue;"><b>Analysis of Missing Data</b></font>

To understand the potential impact of missing values, we can compare the summary statistics of rows with missing data to those without. This comparison will help us determine whether the missing data is completely at random (MCAR) or if there is an underlying pattern causing it.

In [7]:
# filtering through the rows with NaN values in the 'label' column
null_mask = waze_data['label'].isnull()
waze_data_null = waze_data[null_mask]
waze_data_null.describe()

,ID,sessions,drives,total_sessions,n_days_after_onboarding,total_navigations_fav1,total_navigations_fav2,driven_km_drives,duration_minutes_drives,activity_days,driving_days
count,700.000000,700.000000,700.000000,700.000000,700.000000,700.000000,700.000000,700.000000,700.000000,700.000000,700.000000
mean,7405.584286,80.837143,67.798571,198.483348,1709.295714,118.717143,30.371429,3935.967029,1795.123358,15.382857,12.125714
std,4306.900234,79.987440,65.271926,140.561715,1005.306562,156.308140,46.306984,2443.107121,1419.242246,8.772714,7.626373
min,77.000000,0.000000,0.000000,5.582648,16.000000,0.000000,0.000000,290.119811,66.588493,0.000000,0.000000
25%,3744.500000,23.000000,20.000000,94.056340,869.000000,4.000000,0.000000,2119.344818,779.009271,8.000000,6.000000
50%,7443.000000,56.000000,47.500000,177.255925,1650.500000,62.500000,10.000000,3421.156721,1414.966279,15.000000,12.000000
75%,11007.000000,112.250000,94.000000,266.058022,2508.750000,169.250000,43.000000,5166.097373,2443.955404,23.000000,18.000000
max,14993.000000,556.000000,445.000000,1076.879741,3498.000000,1096.000000,352.000000,15135.391280,9746.253023,31.000000,30.000000


In [8]:
# filtering through the rows without NaN values in the 'label' column
not_null_mask = ~waze_data['label'].isnull()
waze_data_not_null = waze_data[not_null_mask]
waze_data_not_null.describe()

,ID,sessions,drives,total_sessions,n_days_after_onboarding,total_navigations_fav1,total_navigations_fav2,driven_km_drives,duration_minutes_drives,activity_days,driving_days
count,14299.000000,14299.000000,14299.000000,14299.000000,14299.000000,14299.000000,14299.000000,14299.000000,14299.000000,14299.000000,14299.000000
mean,7503.573117,80.623820,67.255822,189.547409,1751.822505,121.747395,29.638296,4044.401535,1864.199794,15.544653,12.182530
std,4331.207621,80.736502,65.947295,136.189764,1008.663834,147.713428,45.350890,2504.977970,1448.005047,9.016088,7.833835
min,0.000000,0.000000,0.000000,0.220211,4.000000,0.000000,0.000000,60.441250,18.282082,0.000000,0.000000
25%,3749.500000,23.000000,20.000000,90.457733,878.500000,10.000000,0.000000,2217.319909,840.181344,8.000000,5.000000
50%,7504.000000,56.000000,48.000000,158.718571,1749.000000,71.000000,9.000000,3496.545617,1479.394387,16.000000,12.000000
75%,11257.500000,111.000000,93.000000,253.540450,2627.500000,178.000000,43.000000,5299.972162,2466.928876,23.000000,19.000000
max,14998.000000,743.000000,596.000000,1216.154633,3500.000000,1236.000000,415.000000,21183.401890,15851.727160,31.000000,30.000000


By comparing the two summary statistics, especially focusing on median values and standard deviations, we can see that the results are consistent for both datasets (with and without missing values). Therefore, we can assume that part of the data is missing for random reasons, suggesting that the missing values are likely Missing Completely at Random (MCAR).

<font color='#4393C3' size='5' style="text-decoration: underline; text-decoration-color: blue;"><b>If Missing Data Depends on the User's Device Type</b></font>

We have two types of devices in use: Android and iPhone. If the missing data can be explained by a user's device type, we can use the already filtered waze_data_null dataframe to examine how missing values are distributed between Android and iPhone devices.

In [9]:
# Proportions of the device types in rows with missing labels 
waze_data_null['device'].value_counts(normalize = True).reset_index()

,device,proportion
0,iPhone,0.638571
1,Android,0.361429


In [10]:
# Proportions of the device types in the dataset overall 
waze_data['device'].value_counts(normalize = True).reset_index()

,device,proportion
0,iPhone,0.644843
1,Android,0.355157


We observe approximately the same proportion of devices in the overall dataset and in the dataset with missing 'label' info: 447 iPhone (~ **64%**) vs. 253 (~ **36%**) Android devices. We can assume that the presence of missing values is not dependent on device type.

<font color='#4393C3' size='5' style="text-decoration: underline; text-decoration-color: blue;"><b>Churned vs. Retained Users</b></font>

Now, we can proceed with the analysis of the proportion of churned vs. retained users.

In [11]:
# Proportions of churned and retained uesers in the dataset
waze_data_not_null['label'].value_counts(normalize = True).reset_index()

,label,proportion
0,retained,0.822645
1,churned,0.177355


There is 82% of retained users and 18% churned users in the dataset. 

In [12]:
# Descriptive statistics for the churned users
churned = waze_data_not_null[waze_data_not_null['label'] == 'churned']
churned.describe()

,ID,sessions,drives,total_sessions,n_days_after_onboarding,total_navigations_fav1,total_navigations_fav2,driven_km_drives,duration_minutes_drives,activity_days,driving_days
count,2536.000000,2536.000000,2536.000000,2536.000000,2536.000000,2536.000000,2536.000000,2536.000000,2536.000000,2536.000000,2536.000000
mean,7544.852918,87.238959,72.730678,196.893424,1471.027603,139.414826,31.596609,4147.171864,1975.459630,9.644716,7.218060
std,4401.273381,88.401217,72.082117,140.985386,985.980038,164.007095,48.919135,2499.577849,1465.960239,7.706983,6.543208
min,9.000000,0.000000,0.000000,1.362129,6.000000,0.000000,0.000000,178.232313,23.022685,0.000000,0.000000
25%,3691.250000,25.000000,22.000000,93.496698,626.000000,14.750000,0.000000,2304.118861,928.668633,3.000000,2.000000
50%,7477.500000,59.000000,50.000000,164.339042,1321.000000,84.500000,11.000000,3652.655666,1607.183785,8.000000,6.000000
75%,11364.500000,120.000000,100.000000,265.362864,2238.500000,201.000000,46.000000,5381.361396,2622.819187,14.000000,11.000000
max,14997.000000,743.000000,596.000000,1216.154633,3496.000000,1170.000000,396.000000,19214.475110,10040.568960,31.000000,29.000000


In [13]:
# Descriptive statistics for the retained users
retained = waze_data_not_null[waze_data_not_null['label'] == 'retained']
retained.describe()

,ID,sessions,drives,total_sessions,n_days_after_onboarding,total_navigations_fav1,total_navigations_fav2,driven_km_drives,duration_minutes_drives,activity_days,driving_days
count,11763.000000,11763.000000,11763.000000,11763.000000,11763.000000,11763.000000,11763.000000,11763.000000,11763.000000,11763.000000,11763.000000
mean,7494.673553,79.197654,66.075491,187.963672,1812.359432,117.938451,29.216101,4022.245150,1840.213146,16.816628,13.252827
std,4316.090766,78.918079,64.490746,135.087394,1003.282581,143.682422,44.535100,2505.694427,1443.044031,8.768798,7.675532
min,0.000000,0.000000,0.000000,0.220211,4.000000,0.000000,0.000000,60.441250,18.282082,0.000000,0.000000
25%,3761.500000,23.000000,20.000000,89.891688,962.000000,9.000000,0.000000,2199.661188,826.027498,10.000000,7.000000
50%,7509.000000,56.000000,47.000000,157.586756,1843.000000,68.000000,9.000000,3464.684614,1458.046141,17.000000,14.000000
75%,11241.500000,109.000000,92.000000,250.698741,2685.000000,173.000000,43.000000,5272.026859,2434.386444,24.000000,19.000000
max,14998.000000,725.000000,582.000000,1117.893821,3500.000000,1236.000000,415.000000,21183.401890,15851.727160,31.000000,30.000000


Let's compare the medians (which is better than the mean, because we don’t yet know if our data might be skewed).

In [14]:
# Median values calculation for both 'populations'
waze_data.groupby('label').median(numeric_only=True).reset_index()

,label,ID,sessions,drives,total_sessions,n_days_after_onboarding,total_navigations_fav1,total_navigations_fav2,driven_km_drives,duration_minutes_drives,activity_days,driving_days
0,churned,7477.5,59.0,50.0,164.339042,1321.0,84.5,11.0,3652.655666,1607.183785,8.0,6.0
1,retained,7509.0,56.0,47.0,157.586756,1843.0,68.0,9.0,3464.684614,1458.046141,17.0,14.0


The median values show that churned users have 59 sessions, while retained users have 56 sessions during the last month. We can also see that churned users drove ~200 more kilometers and more that two hours han retained users, even though they have approximately half as much activity (just log-ins) and fewer actual driving days.

In [15]:
# Create a new column 'distance_per_drive' 
waze_data['distance_per_drive'] = waze_data['driven_km_drives'] / waze_data['drives']

# The calculation of the median distance per drive for both 'populations'
waze_data.groupby('label')['distance_per_drive'].median(numeric_only=True).reset_index()

,label,distance_per_drive
0,churned,74.109416
1,retained,75.014702


The median distance per drive for retained users is slightly higher than that for churned users.

In [16]:
# Create a new column 'km_per_driving_day'. Some values in this column will be 'inf' 
# because a drive is only recorded if it exceeds 1 km. If any drive a driver makes  
# over the month is less than 1 km, then the 'drives' column remains 0 while other columns 
# are still updated, leading to division by zero.
waze_data['km_per_driving_day'] = waze_data['driven_km_drives'] / waze_data['driving_days']

# The calculation of the median distance driven per day for both 'populations'
waze_data.groupby('label')['km_per_driving_day'].median(numeric_only=True).reset_index()

,label,km_per_driving_day
0,churned,697.541999
1,retained,289.549333


In [17]:
# How many drives per day each group made
waze_data['drives_per_day'] = waze_data['drives']/waze_data['driving_days']

# Groupping and calculate the median value
waze_data.groupby('label')['drives_per_day'].median(numeric_only = True).reset_index()

,label,drives_per_day
0,churned,10.0000
1,retained,4.0625


From our calculations, we can see that an average churned driver drove approximately 698 km per day, which is about 2.4 times more than an average retained user. Moreover, churned users made 10 drives per day on average, while retained users made around 4. We can assume that this disproportion is due to churned users being long-distance truck drivers (or in a similar profession), and perhaps the app does not meet their needs, leading them to stop using it.

<font color='#4393C3' size='5' style="text-decoration: underline; text-decoration-color: blue;"><b>Is the Proportion of Churned vs. Retained Users Related to Device Type?</b></font>

In [18]:
# Absolute value counts for churned vs. retained users using iPhone or Android
waze_data.groupby(['label','device']).size().reset_index(name='count').sort_values(by='label')

,label,device,count
0,churned,Android,891
1,churned,iPhone,1645
2,retained,Android,4183
3,retained,iPhone,7580


In [19]:
# Absolute value counts for churned vs. retained users using iPhone or Android
waze_data.groupby('label')['device'].value_counts(normalize=True).reset_index(name = 'proportion').sort_values(by='label')

,label,device,proportion
0,churned,iPhone,0.648659
1,churned,Android,0.351341
2,retained,iPhone,0.644393
3,retained,Android,0.355607


The proportion of iPhone to Android users is similar in both the churned and retained groups, and it matches the overall proportion in the entire dataset.

<font color='#4393C3' size='5' style="text-decoration: underline; text-decoration-color: blue;"><b>Conclusions</b></font>

1. The dataset had 700 missing 'label' entries. Analysis of these records suggests that they are most likely missing completely at random (MCAR), with no underlying reason for it.
2. While churned users drove about 2.4 times more per day than retained users, they had roughly half as many driving days. This suggests that the data for churned users may not represent a random sample. The proportion of churned vs. retained users could be skewed because churned users are likely to be long-distance drivers, rather than part of the general population. Improving the app to better meet the needs of this group could help increase retention.
3. Android users accounted for around 36% of the sample, while iPhone users made up about 64%. Data suggests no correlation between the device type and the user status.